libs/motor/inc/public/tfc/motors/positioner.hpp

```diff
@@ -0,0 +1,335 @@
+      : position_update_callback_{ std::move(position_update_callback) } {}
+
+  void first_tacho_update(bool first_new_val) noexcept {
+    if (first_new_val) {
+      increment();
+    }
+
+    auto now{ time_point_t::clock::now() };
+    buffer_.emplace(first_new_val, now);
+  }
+
+  auto increment() noexcept -> void {
+    position_ += 1;
+    if (position_update_callback_) {
+      (*position_update_callback_)(position_);
+    }
+  }
+
+  struct storage {
+    bool tacho_state{};
+    time_point_t time_point{};
+  };
+  circular_buffer<storage, circular_buffer_len> buffer_{};
+  std::int64_t position_{};
+  std::optional<std::function<void(std::int64_t)>> position_update_callback_;
+};
+
+template <typename time_point_t = asio::steady_timer::time_point, std::size_t circular_buffer_len = 1024>
+struct double_tachometer {
+  explicit double_tachometer(std::function<void(std::int64_t)> position_update_callback) noexcept
+      : position_update_callback_{ std::move(position_update_callback) } {}
+
+  // https://github.com/PaulStoffregen/Encoder/blob/master/Encoder.h#L303
+  //                           _______         _______
+  //               Pin1 ______|       |_______|       |______ Pin1
+  // negative <---         _______         _______         __      --> positive
+  //               Pin2 __|       |_______|       |_______|   Pin2
+
+  //	new 	new	    old 	old
+  //	pin2	pin1	pin2	pin1	Result
+  //	----	----	----	----	------
+  //	0	0	0	0	no movement
+  //	0	0	0	1	+1
+  //	0	0	1	0	-1
+  //	0	0	1	1	+2  (assume pin1 edges only)
+  //	0	1	0	0	-1
+  //	0	1	0	1	no movement
+  //	0	1	1	0	-2  (assume pin1 edges only)
+  //	0	1	1	1	+1
+  //	1	0	0	0	+1
+  //	1	0	0	1	-2  (assume pin1 edges only)
+  //	1	0	1	0	no movement
+  //	1	0	1	1	-1
+  //	1	1	0	0	+2  (assume pin1 edges only)
+  //	1	1	0	1	-1
+  //	1	1	1	0	+1
+  //	1	1	1	1	no movement
+  /*
+          // Simple, easy-to-read "documentation" version :-)
+          //
+          void update(void) {
+                  uint8_t s = state & 3;
+                  if (digitalRead(pin1)) s |= 4;
+                  if (digitalRead(pin2)) s |= 8;
+                  switch (s) {
+                          case 0: case 5: case 10: case 15:
+                                  break;
+                          case 1: case 7: case 8: case 14:
+                                  position++; break;
+                          case 2: case 4: case 11: case 13:
+                                  position--; break;
+                          case 3: case 12:
+                                  position += 2; break;
+                          default:
+                                  position -= 2; break;
+                  }
+                  state = (s >> 2);
+          }
+  */
+
+  void first_tacho_update(bool first_new_val) noexcept {
+    if (buffer_.front().first_tacho_state) {
+      if (buffer_.front().second_tacho_state) {
+        if (!first_new_val) {
```

Comment:
I would not consider falling edge reliable source of information, if you want to keep it a duty cycle parameter needs to be added.

Reply:
Duty cycle can be skewed, therefore, each increment can mean different length.

Reply:
the length of each increment is configurable throught `displacement_per_pulse_` variable

Reply:
time of high and low state is not strictly same, but let's not make this stop the PR. The error would be at maximum less than displacement per pulse.
When we need high resolution placement we could implement timers between edges and increase the resolution with that way, then the duty cycle would be determined on the fly.